torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp

```diff
@@ -298,7 +305,8 @@ static void UpdateScalarTypeForInputs(
         input_tensor_type ? input_tensor_type->scalarType() : c10::nullopt;
 
     if ((input->node()->kind() == onnx::Constant) ||
-        (input_scalar_type && (*input_scalar_type != scalar_type))) {
+        (input_scalar_type && (*input_scalar_type != scalar_type) &&
+         !(IsSelectorOp(n->kind())))) {
```

Comment:
The issue with this approach is that it would lead to incorrect behavior when 'condition' input is constant, or when 'x' or 'y' input is non-constant.

Reply:
Hi @BowenBao , thank you very much for reviewing the code.

Indeed the `condition input` must not be considered when walking trough the
constants in an onnx::Where.  How about skipping the first input in case of onnx::where?
I mean something along those lines:

```
diff --git a/torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp b/torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp
index 682798040eb..7de5bd1747a 100644
--- a/torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp
+++ b/torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp
@@ -299,11 +299,17 @@ static void UpdateScalarTypeForInputs(
     return;
   }
 
+  size_t input_idx = 0;
   for (auto input : n->inputs()) {
     auto input_tensor_type = input->type()->cast<TensorType>();
     auto input_scalar_type =
         input_tensor_type ? input_tensor_type->scalarType() : c10::nullopt;
 
+    if (IsSelectorOp(n->kind()) && input_idx == 0) {
+      input_idx++;
+      continue;
+    }
+
     if ((input->node()->kind() == onnx::Constant) ||
         (input_scalar_type && (*input_scalar_type != scalar_type) &&
          !(IsSelectorOp(n->kind())))) {
@@ -330,6 +336,7 @@ static void UpdateScalarTypeForInputs(
         n->replaceInputWith(input, cast_node->output());
       }
     }
+    input_idx++;
   }
 }
```

Regarding the second point you raised (i.e., or when 'x' or 'y' input is
non-constant.),  I must miss something as I do not understand what is the
incorrect behavior I am running into.

This is the current behavior of torch with the `torch.where` and constants.

```python
import torch

c = torch.ones(8, dtype=torch.bool)
x = torch.ones(8, dtype=torch.float16)
y = torch.ones(8, dtype=torch.float32)

out = torch.where(c, x, y)
assert out.dtype == torch.float32

# When a constant is specified as input x or y, the resulting type is the type
# of the non constant. So the output dtype is torch.float16
out = torch.where(c, x, torch.ones(size=(), dtype=torch.float64))
assert out.dtype == torch.float16

# When the input x and y are both constants the output type is the winner type
out = torch.where(
    c,
    torch.ones(size=(), dtype=torch.int64),
    torch.ones(size=(), dtype=torch.float64),
)
assert out.dtype == torch.float64
```

I believe that the changes I introduced are in line with those behaviors.
What am I missing?

Reply:
Thanks, the `index` approach should work. Please add a few comments explaining why first input should be skipped.

Regarding the second point, it looks like the condition ` && !(IsSelectorOp(n->kind()))` you added will skip the following casting section for all where op inputs, unless the first half of the condition `input->node()->kind() == onnx::Constant` eval to true. 

Please note that `constant` and `scalar` are two different concepts. In `torch.jit`, outputs from node of `prim::Constant` or `onnx::Constant` are considered `constant`. `constant` does not imply the value is a `scalar`. Both python number and rank 0 tensor constant are considered `scalar`, otherwise it is considered as `tensor`.

Hence my comment on 'non-constant' refers to values from nodes otherthan `onnx::Constant` or `prim::Constant`. A repro would be like
```python
class Module(nn.Module):
    def forward(self, cond, x, y):
        return torch.where(cond, x, y)
x = torch.randn(3, 2)
y = torch.ones(3, 2).to(torch.int64)
cond = x > 0
```

With added `index` approach you can remove ` && !(IsSelectorOp(n->kind()))` and it should work for all cases.

Reply:
To be extra safe, you should also skip the first input similarly in `InferExpectedScalarType`.

Reply:
Ah ok, now I understand what you meant. Thank you.
Actually, I did introduce the ` && !(IsSelectorOp(n->kind()))` exactly not to touch anything else then the constants (as written in the commit message). But indeed we can fix all the cases where there are no constants at all (and update the final commit message accordingly).

I pushed two more commits on top that address the `InferExpectedScalarType` suggestion and the other comments above.